### PR TITLE
[#187222318] added API for streaming file contents from runs

### PR DESCRIPTION
**Pull request by patmagee**
- Added a new endpoint: `/ga4gh/wes/v1/runs/{run_id}/file` which returns metadata of a file or the bytes of the file depending on the `Accept` header

### CODE REVIEW

File: src/main/java/com/dnastack/wes/files/RunFile.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+@Getter
+@Setter
+@AllArgsConstructor
+@NoArgsConstructor
+@EqualsAndHashCode
+@ToString
+@Builder
+public class RunFile {
+
+    @JsonProperty(value = "file_type")
```
Comment (mpanik) [inline src/main/java/com/dnastack/wes/files/RunFile.java:17]: Let's just use `@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)` on a class

File: src/main/java/com/dnastack/wes/files/RunFileController.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+@RestController
+@RequestMapping("/ga4gh/wes/v1")
+public class RunFileController {
+
+
+    private final RunFileService runFileService;
+
+    public RunFileController(RunFileService runFileService) {this.runFileService = runFileService;}
+
+    @AuditActionUri("wes:run:files:list")
```
Comment (mpanik) [inline src/main/java/com/dnastack/wes/files/RunFileController.java:24]: I always forget what is the standard here, shouldn't it be `wes:runs:files:list`? most right is an action, rest is a resource. Similar to REST API.

File: src/main/java/com/dnastack/wes/files/RunFileController.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    @AuditActionUri("wes:run:files:get-metadata")
+    @PreAuthorize("@accessEvaluator.canAccessResource('/ga4gh/wes/v1/runs/' + #runId, 'wes:runs:read', 'wes')")
+    @GetMapping(value = "/runs/{runId}/file", produces = { MediaType.APPLICATION_JSON_VALUE })
+    public RunFile getRunFile(@PathVariable String runId, @RequestParam String path) {
+        return runFileService.getRunFile(runId,path);
+    }
+
+    @AuditActionUri("wes:run:files:get-metadata")
```
Comment (mpanik) [inline src/main/java/com/dnastack/wes/files/RunFileController.java:38]: is this metadata or get-content?

File: src/main/java/com/dnastack/wes/files/RunFileService.java [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+        if (inputs != null && !inputs.isEmpty()) {
+            inputs.values().forEach(input -> extractFilesFromValue(inputFileSet, OBJECT_MAPPER.valueToTree(input)));
+        }
+
+        extractSecondaryLogFiles(secondaryFileSet, logFileSet, metadataResponse);
+
+        finalFileSet.forEach(path -> files.add(new RunFile(RunFile.FileType.FINAL, path)));
+
+        inputFileSet.forEach(path -> {
+            if (!finalFileSet.contains(path)) {
```
Comment (mpanik) [inline src/main/java/com/dnastack/wes/files/RunFileService.java:82]: let's add this as a Stream API filter instead.

File: src/main/java/com/dnastack/wes/files/RunFileService.java [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+                files.add(new RunFile(RunFile.FileType.INPUT, path));
+            }
+        });
+
+        secondaryFileSet.forEach(path -> {
+            if (!finalFileSet.contains(path) && !logFileSet.contains(path) && !inputFileSet.contains(path)) {
```
Comment (mpanik) [inline src/main/java/com/dnastack/wes/files/RunFileService.java:88]: Ditto

File: src/main/java/com/dnastack/wes/files/RunFileService.java [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+                files.add(new RunFile(RunFile.FileType.SECONDARY, path));
+            }
+        });
+
+        logFileSet.forEach(path -> {
+            if (!finalFileSet.contains(path)) {
```
Comment (mpanik) [inline src/main/java/com/dnastack/wes/files/RunFileService.java:94]: Same

File: src/main/java/com/dnastack/wes/files/RunFileController.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+
+
+    @AuditActionUri("wes:run:files:delete")
+    @PreAuthorize("@accessEvaluator.canAccessResource('/ga4gh/wes/v1/runs/' + #runId, 'wes:runs:write', 'wes')")
+    @DeleteMapping(value = "/runs/{run_id}/files", produces = { MediaType.APPLICATION_JSON_VALUE })
+    public RunFileDeletions deleteRunFiles(
+        @PathVariable("run_id") String runId,
+        @RequestParam(value = "async", required = false) boolean async
```
Comment (mpanik) [inline src/main/java/com/dnastack/wes/files/RunFileController.java:57]: Wdyt of making this `true` by default?

File: src/main/java/com/dnastack/wes/files/RunFileService.java [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+        RunFile runFile = getRunFile(runId, path);
+        long contentLength = runFile.getBlobMetadata().getSize();
+        if (httpRange != null) {
+            long start = httpRange.getRangeStart(runFile.getBlobMetadata().getSize());
+            long end = httpRange.getRangeEnd(runFile.getBlobMetadata().getSize());
+            contentLength = end - start + 1;
+            if (contentLength != runFile.getBlobMetadata().getSize()) {
+                response.setStatus(HttpServletResponse.SC_PARTIAL_CONTENT);
+            }
+        }
```
Comment (mpanik) [inline src/main/java/com/dnastack/wes/files/RunFileService.java:179]: You should be adding Content-Range header as well if range is defined
```
response.setHeader("Content-Range", "bytes " + start + "-" + end + "/" + contentLength);
```

File: src/main/java/com/dnastack/wes/files/RunFileService.java [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+        storageClient.readBytes(response.getOutputStream(), runFile.getPath(), httpRange);
+    }
+
+    private void extractSecondaryLogFiles(Set<String> secondaryFileSet, Set<String> logFileSet, CromwellMetadataResponse metadataResponse) {
+        Map<String, Object> outputs = metadataResponse.getOutputs();
+        if (outputs != null && !outputs.isEmpty()) {
+            outputs.values().forEach(output -> extractFilesFromValue(secondaryFileSet, OBJECT_MAPPER.valueToTree(output)));
+        }
+        Map<String, List<CromwellTaskCall>> calls = metadataResponse.getCalls();
```
Comment (mpanik) [inline src/main/java/com/dnastack/wes/files/RunFileService.java:193]: Should this be hardcoded to Cromwell?

File: src/main/java/com/dnastack/wes/files/RunFileService.java [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+            outputs.values().forEach(output -> extractFilesFromValue(secondaryFileSet, OBJECT_MAPPER.valueToTree(output)));
+        }
+        String stderr = call.getStderr();
+        String stdout = call.getStdout();
+        if (stderr != null && storageClient.isFile(stderr)) {
+            logFileSet.add(stderr);
+        }
+        if (stdout != null && storageClient.isFile(stdout)) {
+            logFileSet.add(stdout);
+        }
```
Comment (mpanik) [inline src/main/java/com/dnastack/wes/files/RunFileService.java:211]:
```java
    Stream.of(call.getStderr(), call.getStdout())
        .filter(Objects::nonNull)
        .filter(storageClient::isFile)
        .forEach(logFileSet::add);
```

**Review by mpanik — COMMENTED** [review]
Looks good! I left few comments

File: src/main/java/com/dnastack/wes/files/RunFileController.java (hunk 7 above)
Comment (patmagee) [reply src/main/java/com/dnastack/wes/files/RunFileController.java:57]: We can make a ticket for that, I would have to look into the behaviour expected in the ewes-service. But I am open to that idea!

File: src/main/java/com/dnastack/wes/files/RunFileService.java (hunk 9 above)
Comment (patmagee) [reply src/main/java/com/dnastack/wes/files/RunFileService.java:193]: The WES service only deals with cromwell at this point so think its okay

File: src/main/java/com/dnastack/wes/files/RunFileController.java [hunk 11]
```diff
@@ -21,21 +21,21 @@ public class RunFileController {
...
-    @AuditActionUri("wes:run:files:get-metadata")
+    @AuditActionUri("wes:runs:files:get-metadata")
     @PreAuthorize("@accessEvaluator.canAccessResource('/ga4gh/wes/v1/runs/' + #runId, 'wes:runs:read', 'wes')")
     @GetMapping(value = "/runs/{runId}/file", produces = { MediaType.APPLICATION_JSON_VALUE })
     public RunFile getRunFile(@PathVariable String runId, @RequestParam String path) {
         return runFileService.getRunFile(runId,path);
     }
 
-    @AuditActionUri("wes:run:files:get-metadata")
+    @AuditActionUri("wes:run:files:get-content")
```
Comment (mpanik) [inline src/main/java/com/dnastack/wes/files/RunFileController.java:38]: Should be `wes:runs:files:get-content`

**Review by mpanik — APPROVED** [review]
@patmagee I left one comment, otherwise it looks good to me!

**Pull request merged**